gobblin-api/src/main/java/org/apache/gobblin/util/io/GsonInterfaceAdapter.java

```diff
@@ -189,7 +192,38 @@ private <S> S readValue(JsonObject jsonObject, TypeToken<S> defaultTypeToken) th
   }
 
   public static <T> Gson getGson(Class<T> clazz) {
-    Gson gson = new GsonBuilder().registerTypeAdapterFactory(new GsonInterfaceAdapter(clazz)).create();
+    Gson gson = new GsonBuilder()
+        .setObjectToNumberStrategy(CustomToNumberPolicy.INTEGER_OR_LONG_OR_DOUBLE)
+        .registerTypeAdapterFactory(new GsonInterfaceAdapter(clazz))
+        .create();
     return gson;
   }
+
+  public enum CustomToNumberPolicy implements ToNumberStrategy {
+    INTEGER_OR_LONG_OR_DOUBLE {
+      @Override
+      public Number readNumber(JsonReader in) throws IOException {
+        String value = in.nextString();
+        try {
+          return Integer.parseInt(value);
+        } catch (NumberFormatException var3) {
+          try {
+            return Long.parseLong(value);
+          } catch (NumberFormatException var2) {
```

Comment:
Nice work on the custom number parsing. But this includes a lot of nesting which could imapact readibility and maintainability , I sugges tbreaking down the nested try-catch blocks in readNumber into smaller methods for each number type (e.g., tryParseInteger, tryParseLong, tryParseDouble). This will improve readability and maintainability by reducing nesting.
Sample code`private Number parseNumber(String value, JsonReader in) throws IOException {
    Number number = tryParseInteger(value);
    if (number != null) return number;

    number = tryParseLong(value);
    if (number != null) return number;

    return tryParseDouble(value, in);
}

private Number tryParseInteger(String value) { /* ... */ }
private Number tryParseLong(String value) { /* ... */ }
private Number tryParseDouble(String value, JsonReader in) throws IOException { /* ... */ }
`